python/convert.c

```diff
@@ -0,0 +1,327 @@
+  switch (upb_fielddef_type(f)) {
+    case UPB_TYPE_ENUM:
+    case UPB_TYPE_INT32:
+      return PyLong_FromLong(val.int32_val);
+    case UPB_TYPE_INT64:
+      return PyLong_FromLongLong(val.int64_val);
+    case UPB_TYPE_UINT32:
+      return PyLong_FromSize_t(val.uint32_val);
+    case UPB_TYPE_UINT64:
+      return PyLong_FromUnsignedLongLong(val.uint64_val);
+    case UPB_TYPE_FLOAT:
+      return PyFloat_FromDouble(val.float_val);
+    case UPB_TYPE_DOUBLE:
+      return PyFloat_FromDouble(val.double_val);
+    case UPB_TYPE_BOOL:
+      return PyBool_FromLong(val.bool_val);
+    case UPB_TYPE_BYTES:
+      return PyBytes_FromStringAndSize(val.str_val.data, val.str_val.size);
+    case UPB_TYPE_STRING:
+      return PyUnicode_DecodeUTF8(val.str_val.data, val.str_val.size, NULL);
+    case UPB_TYPE_MESSAGE:
+      PyErr_Format(PyExc_NotImplementedError,
+                   "Conversion of message types not yet implemented");
+    default:
+      PyErr_Format(PyExc_SystemError,
+                   "Getting a value from a field of unknown type %d",
+                   upb_fielddef_type(f));
+      return NULL;
+  }
+}
+
+static bool PyUpb_GetInt64(PyObject *obj, int64_t *val) {
+  // If the value is already a Python long, retrieves it.  Otherwise performs
+  // any automatic conversions to long (using __index__() or __int__()) that
+  // users expect.
+  *val = PyLong_AsLongLong(obj);
+  if (!PyErr_Occurred()) return true;
+  if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
+    // Rewrite OverflowError -> ValueError.
+    // But don't rewrite other errors such as TypeError.
+    PyErr_Clear();
+    PyErr_Format(PyExc_ValueError, "Value out of range: %S", obj);
+  }
+  return false;
+}
+
+static bool PyUpb_GetUint64(PyObject *obj, uint64_t *val) {
+  // For uint64 Python does not offer any functions as convenient as
+  // PyLong_AsLongLong().  If the object is not already a "long" we must
+  // manually perform the automatic conversion (using __index__() or __int__())
+  // that users expect.
+  if (PyLong_Check(obj)) {
+    *val = PyLong_AsUnsignedLongLong(obj);
+  } else {
+    PyObject* casted = PyNumber_Long(obj);
+    if (!casted) return false;
+    *val = PyLong_AsUnsignedLongLong(casted);
+    Py_DECREF(casted);
+  }
+  if (!PyErr_Occurred()) return true;
+  PyErr_Clear();
+  PyErr_Format(PyExc_ValueError, "Value out of range: %S", obj);
+  return false;
+}
+
+// If `arena` is specified, copies the string data into the given arena.
+// Otherwise aliases the given data.
+static upb_msgval PyUpb_MaybeCopyString(const char *ptr, size_t size,
+                                        upb_arena *arena) {
+  upb_msgval ret;
+  ret.str_val.size = size;
+  if (arena) {
+    char *buf = upb_arena_malloc(arena, size);
+    memcpy(buf, ptr, size);
+    ret.str_val.data = buf;
+  } else {
+    ret.str_val.data = ptr;
+  }
+  return ret;
+}
+
+#define CHECK_RANGE(in, out, low, high) \
+  ((out) = (in), ((out) >= (low) && (out) <= (high)))
```

Comment:
I really don't like having this as a macro.  In particular it evaluates out 3 different times, which is very dangerous

Reply:
Done.